src/objects/jolt_soft_body_impl_3d.cpp

```diff
@@ -0,0 +1,744 @@
+		ERR_FAIL_COND_D(mesh_data.is_empty());
+
+		const PackedInt32Array mesh_indices = mesh_data[RenderingServer::ARRAY_INDEX];
+		ERR_FAIL_COND_D(mesh_indices.is_empty());
+
+		const PackedVector3Array mesh_vertices = mesh_data[RenderingServer::ARRAY_VERTEX];
+		ERR_FAIL_COND_D(mesh_vertices.is_empty());
+
+		iter_shared_data = mesh_to_shared.emplace(mesh);
+
+		LocalVector<int32_t>& mesh_to_physics = iter_shared_data->second.mesh_to_physics;
+
+		JPH::SoftBodySharedSettings& settings = *iter_shared_data->second.settings;
+		settings.mVertexRadius = JoltProjectSettings::get_soft_body_point_margin();
+
+		JPH::Array<SoftBodyVertex>& physics_vertices = settings.mVertices;
+		JPH::Array<SoftBodyFace>& physics_faces = settings.mFaces;
+		JPH::Array<SoftBodyEdge>& physics_edges = settings.mEdgeConstraints;
+
+		HashMap<Vector3, int32_t> vertex_to_physics;
+
+		const auto mesh_vertex_count = (int32_t)mesh_vertices.size();
+		const auto mesh_index_count = (int32_t)mesh_indices.size();
+
+		mesh_to_physics.resize(mesh_vertex_count);
+		physics_vertices.reserve((size_t)mesh_vertex_count);
+		vertex_to_physics.reserve(mesh_vertex_count);
+
+		int32_t physics_index_count = 0;
+
+		auto is_face_degenerate = [](const int32_t p_face[3]) {
+			return p_face[0] == p_face[1] || p_face[0] == p_face[2] || p_face[1] == p_face[2];
+		};
+
+		for (int32_t i = 0; i < mesh_index_count; i += 3) {
+			int32_t physics_face[3];
+			int32_t mesh_face[3];
+
+			for (int32_t j = 0; j < 3; ++j) {
+				const int32_t mesh_index = mesh_indices[i + j];
+				const Vector3 vertex = mesh_vertices[mesh_index];
+
+				auto iter_physics_index = vertex_to_physics.find(vertex);
+
+				if (iter_physics_index == vertex_to_physics.end()) {
+					physics_vertices.emplace_back(JPH::Float3(vertex.x, vertex.y, vertex.z));
+					iter_physics_index = vertex_to_physics.insert(vertex, physics_index_count++);
+				}
+
+				mesh_face[j] = mesh_index;
+				physics_face[j] = iter_physics_index->second;
+				mesh_to_physics[mesh_index] = iter_physics_index->second;
+			}
+
+			ERR_CONTINUE_MSG(
+				is_face_degenerate(physics_face),
+				vformat(
+					"Failed to append face to soft body '%s'. "
+					"Face was found to be degenerate. "
+					"Face consist of indices %d, %d and %d.",
+					to_string(),
+					mesh_face[0],
+					mesh_face[1],
+					mesh_face[2]
+				)
+			);
+
+			// Jolt uses a different winding order, so we swap the indices to account for that.
+
+			physics_faces.emplace_back(
+				(JPH::uint32)physics_face[2],
+				(JPH::uint32)physics_face[1],
+				(JPH::uint32)physics_face[0]
+			);
+		}
+
+		// HACK(mihe): Since Godot's stiffness is input as a coefficient between 0 and 1, and Jolt
+		// uses actual stiffness for its edge constraints, we crudely map one to the other with an
+		// arbitrary constant.
+		const float stiffness = MAX(stiffness_coefficient * 100'000.0f, 0.00001f);
+		const float inverse_stiffness = 1.0f / stiffness;
```

Comment:
@jrouwe I'm curious if you have any better idea for how to deal with `mCompliance`.

The `linear_stiffness` property in Godot is sadly not actual stiffness but instead a fraction/coefficient between 0 and 1, and I'm running out of ideas for how to best map from this to actual stiffness, so I just settled for this naive/crude linear mapping.

In Godot Physics this particular property does nothing, as it turns out, as it's only ever incorporated into one of the "link constants" (see [here](https://github.com/godotengine/godot/blob/9adb7c7d130c6d4eb0e80b92d6eebd71fac3384d/servers/physics_3d/godot_soft_body_3d.cpp#L280)) but that constant is never actually used anywhere.

In Bullet however, from where I'm assuming this was copied, this "link constant" does do something, but I'm struggling to figure out exactly how it's different from Jolt, and whether there's a way to map between the two. It doesn't seem to be affected by mass at all, from what I can tell, as the constraints seem to act about the same no matter how much mass I load on.

Given that this property isn't actually used in Godot Physics I wouldn't be entirely against just not letting this be configurable (which would also make the caching of the shared soft body settings easier) and instead either set it to be some constant or calculate it automatically somehow. I tried a couple of variations based on mass, gravity and the rest length of the edge, but I couldn't really find anything that made much sense, nor felt intuitive.

Maybe just hardcoding it to `0.000001f` is the better play?

Reply:
Are you sure it isn't used? To me it looks like it is used here:

https://github.com/godotengine/godot/blob/9adb7c7d130c6d4eb0e80b92d6eebd71fac3384d/servers/physics_3d/godot_soft_body_3d.cpp#L1104

(it is used in link.c2 as well but that constant doesn't seem to be used)

Reply:
You're absolutely right. I'm not sure why I stopped looking after the first occurrence. That's pretty much the same as with Bullet then: https://github.com/godotengine/godot/blob/44cb43b5e73a4b40ae20cc5d434d3e70cb0672c0/thirdparty/bullet/BulletSoftBody/btSoftBody.cpp#L4009

That makes the hardcoded value less appealing I guess.

Either way, would you happen to have any suggestions for how to map one to the other?

Reply:
If I compare the formulas in Bullet and GodotJolt then I think they're pretty much the same. The update of an edge constraint boils down to:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/afe910bb-395e-42e5-a854-dc08da085e20)

(and something similar for delta_p2)

Where p1 and p2 are the positions of the particles, m1 and m2 the masses, rest_len is the rest length between the particles, len the current length (|p2 - p1|). 

This formula strikes me as very odd since rest_len^2 doesn't have the same units as len so you cannot subtract them. I don't recognize the formula, I would expect something like eq 10 of https://matthias-research.github.io/pages/publications/posBasedDyn.pdf.

If eq 10 was the intention then a linear_stiffness (with which we would multiply the right hand side of eq 10) between 0 and 1 starts making sense: If it is 0 then nothing of the position is corrected, if it is 1 then the edge length is fully updated to rest_len. So a linear_stiffness of 1 would map to a compliance of 0 and a linear stiffness of 0 would map to a compliance of infinite. Problem is that godot isn't using eq 10 so I'm not sure what to make of this.

Reply:
For reference, this is what happens in Jolt with XPBD:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/764bc41c-f27a-4c47-a83b-3d585d2543fb)

Reply:
Interesting. Thank you for the breakdown/analysis.

I'm getting the sense that this mapping will end up being somewhat arbitrary in that case. I might take some creative liberties and make it a cubic mapping then. That way it should be possible to make things loose at the lower ranges with a low mass, since the increments are 0.01, while still being able to really crank it up if need be.

Reply:
On a somewhat related note, is there any chance that there could be some sort of per-body override of `mCompliance`, so that I can avoid having to use the per-edge `mCompliance` altogether? With the way things are now in Jolt, I don't see any good way of actually changing compliance after the body has been created, which is something you can do in Godot.

Recreating the body through `GetSoftBodyCreationSettings` doesn't seem to set new initial positions of all the vertices, since it just copies the settings pointer, which I guess makes sense, but that also means you can't really adjust the compliance without a lot of jank. Even if `GetSoftBodyCreationSettings`/`SoftBodySharedSettings` did support this, I would need to cache the shared settings based on not only the mesh, but stiffness as well, which becomes somewhat wasteful, given that this stiffness applies to the entire body in Godot.

Reply:
> On a somewhat related note, is there any chance that there could be some sort of per-body override of mCompliance, so that I can avoid having to use the per-edge mCompliance altogether?

I would prefer not to do this. This would basically mean that I need to duplicate this code:

https://github.com/jrouwe/JoltPhysics/blob/a6c9f8808f78c760ed0bed6f2e6c6129e66e9203/Jolt/Physics/SoftBody/SoftBodyMotionProperties.cpp#L287-L303

with the only difference that ```e.mCompliance``` has a different value. A conditional in that place is not a good idea since it is the innermost loop and executed many thousand times per frame.

I don't think you need to recreate the body if the compliance changes. I'm assuming you create a single SoftBodySharedSettings object per soft body since I don't think godot shares anything between the soft bodies (but maybe I'm wrong considering your remark above). I think you can just loop over the edges and update the compliance and the soft body will automatically use the new values.

Reply:
> I would prefer not to do this.

Fair enough. That makes sense. I figured I should check at least.

> I'm assuming you create a single SoftBodySharedSettings object per soft body since I don't think godot shares anything between the soft bodies (but maybe I'm wrong considering your remark above).

Godot does in fact share resources quite aggressively. If you duplicate a `SoftBody3D` you'll end up using the same `Mesh` resource for both of them, as well as whatever other resources it might reference. It seemed like a shame not to utilize this trait and cache Jolt's `SoftBodySharedSettings` based on this mesh resource ID.

I guess the question becomes whether the caching/sharing is worth more than the ability to change `linear_stiffness` after the `SoftBody3D` has been assigned a mesh, which is when the underlying Jolt body gets created. I have a hard time imagining that many people would mess with `linear_stiffness` in any sort of dynamic fashion, whereas many more people might benefit from the performance/memory benefits that the caching brings with it.

I'm also fairly certain that you can't actually change `linear_stiffness` dynamically in Godot Physics either, or in Bullet in Godot 3 for that matter, as they don't seem to actually update the edge constants after changing `linear_stiffness`.